### PR TITLE
Factories for private Gitlab repository with pre-created personal access token

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
This PR allows to accept factories to private Gitlab repositories using pre-created personal access tokens stored as a secret inside the user namespace. Secret example is provided below:

```
kind: Secret
apiVersion: v1
metadata:
  annotations:
    che.eclipse.org/che-userid: 73a27d90-3d63-4d8c-a8c6-9b4a716c702a
    che.eclipse.org/expired-after: '-1'
    che.eclipse.org/scm-url: 'https://gitlab.apps.cluster-2ab2.2ab2.example.opentlc.com'
    che.eclipse.org/scm-userid: '1'
    che.eclipse.org/scm-username: root
  name: gitlab-personal-access-token-secret
  labels:
    app.kubernetes.io/component: scms-personal-access-token
    app.kubernetes.io/part-of: che.eclipse.org
data:
  token: Yzh5cEt6cURxUWVCa3FKazhtaHg=
type: Opaque
```

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/19170

### How to test this PR?
Prepare private Gitblab repository;
Create and save personal access token on Gitlab side;
Configure Che server to use Gitlab instance and prepare secret in user namespace;
Run factory to prevate repo. Factory should work.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java [hunk 1]
```diff
@@ -185,15 +183,22 @@ public String location() {
...
-    }
-    joiner.add("-").add("raw").add(firstNonNull(branch, "master"));
-    if (subfolder != null) {
-      joiner.add(subfolder);
+    StringJoiner joiner =
+        new StringJoiner("/")
+            .add(hostName)
+            .add("api/v4/projects")
+            .add(username + "%2f" + project)
+            .add("repository")
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java:195]: repository can't be null?

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import org.eclipse.che.api.workspace.server.devfile.exception.DevfileException;
+import org.eclipse.che.commons.env.EnvironmentContext;
+
+
+/**
+ * Common implementation of file content provider which is able to access
+ * content of private repositories using  personal access tokens from specially
+ * formatted secret in user-s namespace.
+ */
+public class AuthorizingFileContentProvider<T extends RemoteFactoryUrl>
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java:33]: Do we need a test for this class? Can we have it?

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java (hunk 1 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java:195]: It's just the API keyword indicating you accessing the repository part of the project

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java (hunk 2 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java:33]: Right now it's tested via both extension classes. But i will consider adding own one.

**Review by skabashnyuk — APPROVED** [review]

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java [hunk 3]
```diff
@@ -185,15 +183,22 @@ public String location() {
...
-      joiner.add(repository);
-    }
-    joiner.add("-").add("raw").add(firstNonNull(branch, "master"));
-    if (subfolder != null) {
-      joiner.add(subfolder);
+    StringJoiner joiner =
+        new StringJoiner("/")
+            .add(hostName)
+            .add("api/v4/projects")
+            .add(username + "%2f" + project)
```
Comment (sparkoo) [inline wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java]: why this needs to be encoded as `%2f` ?

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java [hunk 4]
```diff
@@ -185,15 +183,22 @@ public String location() {
...
+            .add("repository")
+            .add("files")
+            .add(fileName)
+            .add("raw");
+    String resultUrl = joiner.toString();
```
Comment (sparkoo) [inline wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java]: I guess we can do directly `String resultsUrl = new StringJoiner(..)....toString()`, without the extra `joiner` variable

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+import org.eclipse.che.api.factory.server.urlfactory.RemoteFactoryUrl;
+import org.eclipse.che.api.workspace.server.devfile.FileContentProvider;
+import org.eclipse.che.api.workspace.server.devfile.URLFetcher;
+import org.eclipse.che.api.workspace.server.devfile.exception.DevfileException;
+import org.eclipse.che.commons.env.EnvironmentContext;
+
+
+/**
+ * Common implementation of file content provider which is able to access
+ * content of private repositories using  personal access tokens from specially
```
Comment (sparkoo) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java]:
```suggestion
 * content of private repositories using personal access tokens from specially
```

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+ * formatted secret in user-s namespace.
```
Comment (sparkoo) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/scm/AuthorizingFileContentProvider.java]:
```suggestion
 * formatted secret in user's namespace.
```

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/DefaultFactoryUrl.java [hunk 7]
```diff
@@ -40,6 +41,18 @@ public String location() {
         });
   }
 
+  // never really used, just for compatibility
```
Comment (sparkoo) [inline wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/DefaultFactoryUrl.java]: shouldn't it be abstract then? or throw an exception? or return null?

**Review by sparkoo — COMMENTED** [review]
lgtm, few minor comments

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java (hunk 3 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabUrl.java]: It's requirement of API:
image

File: wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/DefaultFactoryUrl.java (hunk 7 above)
Comment (mshaposhnik) [reply wsmaster/che-core-api-factory/src/main/java/org/eclipse/che/api/factory/server/urlfactory/DefaultFactoryUrl.java]: i would left it, as it is properly implemented, just not used now (but someone can use it). So i;''ll just remove those confusing comments. Ok?

**Review by sparkoo — APPROVED** [review]

**Comment by mshaposhnik** [comment]
[crw-ci-test]

**Pull request merged**